Bazelisk is pulling a non existing release ( 3.5.1 ) when invoked w/o .bazelversion

Invoking `bazelisk build` is attempting to pull a non-existent release.

This afternoon, a handful of xchammer github jobs started failing that used bazelisk. For example:
https://github.com/pinterest/xchammer/pull/259/checks?check_run_id=1168118720

The xchammer CI runs `bazelisk` and we don't have a version specified in the repository. Without the version, it attempts to pull a non-existing URL:
```
2020/09/25 21:40:42 Downloading https://releases.bazel.build/3.5.1/release/bazel-3.5.1-darwin-x86_64...
2020/09/25 21:40:42 could not download Bazel: HTTP GET https://releases.bazel.build/3.5.1/release/bazel-3.5.1-darwin-x86_64 failed with error 404
make: *** [build-impl] Error 1
Error: Process completed with exit code 2.```

I checked on Bazel releases this afternoon, and it looks like there is no `3.5.1` bazel release: https://github.com/bazelbuild/bazel/releases

--- a/.bazelci/config.yml
+++ b/.bazelci/config.yml
@@ -3,6 +3,8 @@ platforms:
   ubuntu1604:
     test_targets:
       - //:go_bazelisk_test
+      - //:go_default_test
+      - //:go_version_test
       - //:py_bazelisk_test
       - //:py3_bazelisk_test
     test_flags:
@@ -11,6 +13,8 @@ platforms:
   ubuntu1804:
     test_targets:
       - //:go_bazelisk_test
+      - //:go_default_test
+      - //:go_version_test
       - //:py_bazelisk_test
       - //:py3_bazelisk_test
     test_flags:
@@ -19,6 +23,8 @@ platforms:
   macos:
     test_targets:
       - //:go_bazelisk_test
+      - //:go_default_test
+      - //:go_version_test
       - //:py_bazelisk_test
       - //:py3_bazelisk_test
     test_flags:
@@ -27,6 +33,8 @@ platforms:
   windows:
     test_targets:
       - //:go_bazelisk_test
+      - //:go_default_test
+      - //:go_version_test
       - //:py_bazelisk_test
     test_flags:
       - --flaky_test_attempts=1

--- a/BUILD
+++ b/BUILD
@@ -70,6 +70,10 @@ go_test(
     embed = [":go_default_library"],
     importpath = "github.com/bazelbuild/bazelisk",
     deps = [
+        "//core",
+        "//httputil",
+        "//repositories",
+        "//versions",
         "@io_bazel_rules_go//go/tools/bazel:go_default_library",
     ],
 )

--- a/bazelisk_version_test.go
+++ b/bazelisk_version_test.go
@@ -61,6 +61,32 @@ func TestResolveLatestRcVersion(t *testing.T) {
 	}
 }
 
+func TestResolveLatestVersion_TwoLatestVersionsDoNotHaveAReleaseYet(t *testing.T) {
+	listBody := buildGCSResponseOrFail(t, []string{"4.0.0/", "5.0.0/", "6.0.0/"}, []interface{}{})
+	transport.AddResponse("https://www.googleapis.com/storage/v1/b/bazel/o?delimiter=/", 200, listBody)
+
+	v4ReleaseBucket := buildGCSResponseOrFail(t, []string{}, []interface{}{"fake_release_item__since_this_is_a_release"})
+	transport.AddResponse("https://www.googleapis.com/storage/v1/b/bazel/o?delimiter=/&prefix=4.0.0/release/", 200, v4ReleaseBucket)
+
+	v5ReleaseBucket := buildGCSResponseOrFail(t, []string{}, []interface{}{})
+	transport.AddResponse("https://www.googleapis.com/storage/v1/b/bazel/o?delimiter=/&prefix=5.0.0/release/", 200, v5ReleaseBucket)
+
+	v6ReleaseBucket := buildGCSResponseOrFail(t, []string{}, []interface{}{})
+	transport.AddResponse("https://www.googleapis.com/storage/v1/b/bazel/o?delimiter=/&prefix=6.0.0/release/", 200, v6ReleaseBucket)
+
+	gcs := &repositories.GCSRepo{}
+	repos := core.CreateRepositories(gcs, nil, nil, nil, false)
+	version, _, err := repos.ResolveVersion(tmpDir, versions.BazelUpstream, "latest")
+
+	if err != nil {
+		t.Fatalf("Version resolution failed unexpectedly: %v", err)
+	}
+	expectedVersion := "4.0.0"
+	if version != expectedVersion {
+		t.Fatalf("Expected version %s, but got %s", expectedVersion, version)
+	}
+}
+
 func TestResolveLatestVersion_GCSIsDown(t *testing.T) {
 	transport.AddResponse("https://www.googleapis.com/storage/v1/b/bazel/o?delimiter=/", 500, "")
 

--- a/core/repositories.go
+++ b/core/repositories.go
@@ -1,4 +1,3 @@
-// Package core contains interfaces to work with Bazel repositories.
 package core
 
 import (

--- a/repositories/gcs.go
+++ b/repositories/gcs.go
@@ -49,14 +49,22 @@ func getVersionHistoryFromGCS(onlyFullReleases bool) ([]string, error) {
 	available := getVersionsFromGCSPrefixes(prefixes)
 	sorted := versions.GetInAscendingOrder(available)
 
-	if onlyFullReleases && len(sorted) > 0 {
-		latestVersion := sorted[len(sorted)-1]
-		_, isRelease, err := listDirectoriesInReleaseBucket(latestVersion + "/release/")
-		if err != nil {
-			return []string{}, fmt.Errorf("could not list release candidates for latest release: %v", err)
+	// TODO(#171): This algorithm is incorrect if version == 'latest-n' and any of the last n versions (except the last one) does not have a release yet.
+	if onlyFullReleases {
+		for len(sorted) > 0 {
+			lastIndex := len(sorted)-1
+			latestVersion := sorted[lastIndex]
+			_, isRelease, err := listDirectoriesInReleaseBucket(latestVersion + "/release/")
+			if err != nil {
+				return []string{}, fmt.Errorf("could not list release candidates for latest release: %v", err)
+			}
+			if isRelease {
+				break
+			}
+			sorted = sorted[:lastIndex]
 		}
-		if !isRelease {
-			sorted = sorted[:len(sorted)-1]
+		if len(sorted) == 0 {
+			return []string{}, errors.New("there are no releases available")
 		}
 	}
 

--- a/repositories/gcs.go
+++ b/repositories/gcs.go
@@ -52,11 +52,11 @@ func getVersionHistoryFromGCS(onlyFullReleases bool) ([]string, error) {
 	// TODO(#171): This algorithm is incorrect if version == 'latest-n' and any of the last n versions (except the last one) does not have a release yet.
 	if onlyFullReleases {
 		for len(sorted) > 0 {
-			lastIndex := len(sorted)-1
+			lastIndex := len(sorted) - 1
 			latestVersion := sorted[lastIndex]
 			_, isRelease, err := listDirectoriesInReleaseBucket(latestVersion + "/release/")
 			if err != nil {
-				return []string{}, fmt.Errorf("could not list release candidates for latest release: %v", err)
+				return []string{}, fmt.Errorf("could not list available releases for %v: %v", latestVersion, err)
 			}
 			if isRelease {
 				break

